include/pqxx/util.hxx

```diff
@@ -280,18 +280,135 @@ struct PQXX_LIBEXPORT thread_safety_model
 #  define PQXX_POTENTIAL_BINARY_ARG typename
 #endif
 
+struct byte_char_traits : std::char_traits<char>
+{
+  using char_type = std::byte;
+
+  static void assign(std::byte &a, const std::byte &b) noexcept { a = b; }
+
+  static bool eq(std::byte a, std::byte b) { return a == b; }
+
+  static bool lt(std::byte a, std::byte b) { return a < b; }
+
+  static int compare(const std::byte *a, const std::byte *b, std::size_t size)
+  {
+    for (; size; --size, ++a, ++b)
+    {
+      if (lt(*a, *b))
+        return -1;
+      if (lt(*b, *a))
+        return 1;
+    }
+
+    return 0;
+  }
+
+  static size_t length(const std::byte *data)
+  {
+    std::size_t length{};
+
+    for (; !eq(*data, std::byte{0}); ++data) ++length;
+
+    return length;
+  }
+
+  static const std::byte *
+  find(const std::byte *data, std::size_t size, const std::byte &val)
+  {
+    for (; size; --size, ++data)
+    {
+      if (eq(*data, val))
+        return data;
+    }
+
+    return nullptr;
+  }
+
+  static std::byte *
+  move(std::byte *dst, const std::byte *src, std::size_t size)
+  {
+    if (size == 0)
+      return dst;
+
+    std::byte *beginning{dst};
+
+    if (dst < src)
+    {
+      for (; size; --size, ++dst, ++src) assign(*dst, *src);
+    }
+    else if (src < dst)
+    {
+      dst += size;
+      src += size;
+
+      for (; size; --size) assign(*--dst, *--src);
+    }
+
+    return beginning;
+  }
+
+  static std::byte *
+  copy(std::byte *dst, const std::byte *src, std::size_t size)
+  {
+    std::byte *beginning{dst};
+
+    for (; size; --size, ++dst, ++src) assign(*dst, *src);
+
+    return beginning;
+  }
+
+  static std::byte *assign(std::byte *dst, std::size_t size, std::byte value)
+  {
+    std::byte *beginning{dst};
+
+    for (; size; --size, ++dst) assign(*dst, value);
+
+    return beginning;
+  }
+
+  static int_type not_eof(int_type value)
+  {
+    return eq_int_type(value, eof()) ? ~eof() : value;
+  }
+
+  static std::byte to_char_type(int_type value) { return std::byte(value); }
+
+  static int_type to_int_type(std::byte value) { return int_type(value); }
+
+  static bool eq_int_type(int_type a, int_type b) { return a == b; }
+
+  static int_type eof() { return int_type(EOF); }
+};
+
+template<typename TYPE, typename = void>
+struct has_generic_char_traits : std::false_type
+{};
+
+template<typename TYPE>
+struct has_generic_char_traits<
+  TYPE, std::void_t<decltype(std::char_traits<TYPE>::eof)>> : std::true_type
+{};
+
+inline constexpr bool has_generic_bytes_char_traits =
+  has_generic_char_traits<std::byte>::value;
+using bytes = std::conditional<
```

Comment:
These types _definitely_ need documentation, because we'll expect users to understand them.

In particular...  Do you think it would make sense to put a big warning in there saying "we may suddenly decide to replace this type with `std::vector<std::byte>` in the future, but we're not doing that right now for historical reasons"?